### PR TITLE
adding some higher level methods in settings.Provider

### PR DESCRIPTION
**Pull request by alessandrozucca**

### CODE REVIEW

File: aggregate_provider.go [hunk 1]
```diff
@@ -1,5 +1,11 @@
 package setting
 
+import (
+	"time"
+
+	"github.com/spf13/cast"
```
Comment (alistairjudson) [inline aggregate_provider.go]: Why do you need to pull in an external dependency?

File: provider.go [hunk 2]
```diff
@@ -1,11 +1,24 @@
...
+	"github.com/spf13/cast"
+)
+
 // Provider is a simple interface, that can your types
 // should accept if you want access to raw settings, or that you
 // should implement if you want your structs to be auto-filled
 // from a source of your choice
 type Provider interface {
 	Load(keys ...string) (map[string]Setting, ProviderError)
+	GetString(key string) string
```
Comment (alistairjudson) [inline provider.go]: This interface is not idiomatic at all, the amount of methods on this interface is excessive. https://go-proverbs.github.io/ "The bigger the interface, the weaker the abstraction"

File: provider.go [hunk 3]
```diff
@@ -1,11 +1,24 @@
...
+	GetBool(key string) bool
+	GetInt(key string) int
+	GetInt64(key string) int64
+	GetFloat64(key string) float64
+	GetTime(key string) time.Time
```
Comment (alistairjudson) [inline provider.go]: What possible use is there for loading times + durations from settings?

File: provider.go [hunk 4]
```diff
@@ -23,3 +36,52 @@ type ProviderError interface {
...
+func (pf ProviderFunc) GetString(key string) string {
+	s, _ := pf.Load(key)
+	val := s[key].String()
+	return cast.ToString(val)
+}
+
+// GetBool returns the value associated with the key as a boolean.
+func (pf ProviderFunc) GetBool(key string) bool {
+	s, _ := pf.Load(key)
+	val := s[key].String()
```
Comment (alistairjudson) [inline provider.go]: This will panic if the key is not found.

File: glide.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package: github.com/intelligentpos/setting
+import:
+- package: github.com/spf13/cast
```
Comment (alistairjudson) [inline glide.yaml]: you've pulled in a lot of dependencies here, for no real reason, I understand the use of UUID, but most other things can be achieved by using the stdlib.

File: provider.go (hunk 3 above)
Comment (alessandrozucca) [inline provider.go]: Not sure - I just pulled in the main functions from https://github.com/spf13/viper/blob/master/viper.go#L677

I guess `viper` has `GetTime()` cause needs to handle a lot of use cases..

File: provider.go (hunk 2 above)
Comment (alessandrozucca) [inline provider.go]: Very true, it was a bit of a hack to keep backward compatibility with the code that uses already the Provider.. 

I can split it in more interfaces..

**Comment by alistairjudson** [comment]
I'm closing this pull request, the aim of setting is really to provide a simple way of loading environmental settings into a service, mainly through the use of tags on a struct. 

The provider + setting interfaces are a little clunky for this reason, as they are intended to be used by a filler.

The changes that you are proposing change the very unambiguous states of: 

- Key exists but is unable to be represented in the desired type
- And Key does not exist.

Into just getting the 0 value for that type, at best and possibly a panic at worst.

The general use case for this for loading settings into structs, with the config for other packages, e.g. Database settings, Key/Cert locations, not for loading single settings, especially boolean settings, which may end up littering the codebase with conditionals, giving vastly different behaviour based on boolean settings.

I, and others do not see any reason to add additional functionality to this tool at this time.

You have also dramatically reduced test coverage.

It is for these reasons, and the reasons I have left in comments that I have decided to close this pull request.

**Comment by alessandrozucca** [comment]
Thats ok I understand. 
Basically the package is being used for the wrong purpose if we use it to load single settings. 
Fair enough, thanks for the explanation.

**Pull request closed without merging**